### ISSUE TITLE
Adam optimizer in sPacy giving strange results

### ISSUE THREAD

**Issue by themrmax** [explosion/spaCy#6799]
## How to reproduce the behaviour
[Not sure if this is a bug, or if I'm confused about usage. I've also cross-posted to stackoverflow if it's more appropriate as a question there: https://stackoverflow.com/questions/65865385/adam-optimizer-in-spacy-giving-strange-results]

I have the following toy example to play with the `TextCategorizer` CNN model in sPacy. Given the distribution of my input dataset, I am expecting the test results to converge to 0.75, 0, and 0.5 respectively. However this never happens; rather each result either converges to 1.0 or 0.0, seemingly at random. Can anyone help explain to me what is happening here and if I can fix my model to make it deterministic?

Note: changing the architecture to `"bow"` (bag of words) does give the expected results, so I guess the issue is somehow related to the CNN layers ...

Example code:

```
import spacy

num_epochs = 100
drop = 0.0
learn_rate = 0.01
architecture = "simple_cnn"

model = spacy.load('en_core_web_sm')

model.add_pipe(model.create_pipe("textcat", config={"architecture": architecture}))
model.get_pipe('textcat').add_label("green")

data = [('apple 1', {"cats": {"green": 1.0}}),
        ('apple 2', {"cats": {"green": 1.0}}),
        ('apple 3', {"cats": {"green": 1.0}}),
        ('apple 4', {"cats": {"green": 0.0}}),
        ('banana 1', {"cats": {"green": 0.0}}),
        ('banana 2', {"cats": {"green": 0.0}}),
        ('banana 3', {"cats": {"green": 0.0}}),
        ('banana 4', {"cats": {"green": 0.0}}),
        ('dz1909 1', {"cats": {"green": 0.0}}),
        ('dz1909 2', {"cats": {"green": 0.0}}),
        ('dz1909 3', {"cats": {"green": 1.0}}),
        ('dz1909 4', {"cats": {"green": 1.0}}),
        ]


sgd = model.begin_training()
sgd.learn_rate = learn_rate
for i in range(num_epochs):
    print("epoch: ", i)
    print("apple", model("apple").cats, "should converge to 0.75")
    print("banana", model("banana").cats, "should converge to 0")
    print("dz1909", model("dz1909").cats, "should converge to 0.5")
    for doc, annot in data:
        model.update([doc], [annot], drop=drop, sgd=sgd)
```
## Your Environment

* **spaCy version:** 2.3.5
* **Platform:** Darwin-20.2.0-x86_64-i386-64bit
* **Python version:** 3.7.4

**Comment by honnibal** [comment]
It's definitely possible that there are some bugs in ML functions in v2 that could show up on smaller tests even if larger datasets work fine. However, the behaviour you're seeing might not actually be a bug (which is what makes this hard).

I'm not sure how much detail to go into here, but any of the following could be problems:

1. The random seed isn't being set, and some random seeds won't converge correctly
2. The learning rate might be too high
3. The data isn't being minibatched or shuffled, so the model can get stuck in the same cycle of updates
4. There's L2 regularization by default, which might be too high
5. The Adam default parameters do lead to some models getting "stuck"

Fundamentally, the neural network models start at a disadvantage from bag-of-words because they have to learn a distributed representation to even recognise different words. The embedding representation in spaCy is relatively intricate, consisting of several hash embeddings of different subfeatures of the words. If you have a combination of circumstances that let the model get stuck, it might not be able to learn to recognise the words, and then never make progress. The bag-of-words architecture won't have this problem.
